### PR TITLE
Fix Numpy on PyPy

### LINKED ISSUE #53752: Enable numpy on pypy
Numpy now supports pypy. However, one of it's dependencies, funcsigs, has it's test runner broken on pypy3 (the package itself works). Please use Gentoo's patch for funcsigs on pypy3 and reenable pypy for numpy.

### PR DESCRIPTION
**Pull request by leo60228**
###### Motivation for this change
Numpy works on PyPy as of October 2017, but tests fail for one of its dependencies. I applied a patch from Gentoo, disabled tests on pytest for PyPy (due to an upstream bug), and disabled tests on Numpy for PyPy (since documentation claims that they fail and I'm too impatient to test).

Resolves #53752

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by domenkozar** [comment]
@GrahamcOfBorg build pypyPackages.numpy

File: pkgs/development/python-modules/funcsigs/default.nix [hunk 1]
```diff
@@ -9,6 +9,10 @@ buildPythonPackage rec {
     inherit pname version;
     sha256 = "0l4g5818ffyfmfs1a924811azhjj8ax9xd1cffr1mzd3ycn0zfx7";
   };
+  
+  patches = [
+    ./funcsigs-1.0.2-fix-pypy3-tests.patch
```
Comment (veprbl) [inline pkgs/development/python-modules/funcsigs/default.nix]: Let's use `fetchpatch` to download the patch from https://raw.githubusercontent.com/gentoo/gentoo/fd9a5a0a3c15d7400b936ce0f0814fd4078e900c/dev-python/funcsigs/files/funcsigs-1.0.2-fix-pypy3-tests.patch

Comment (leo60228) [reply pkgs/development/python-modules/funcsigs/default.nix]: How do I do that?

File: pkgs/development/python-modules/pytest/default.nix [hunk 2]
```diff
@@ -23,7 +23,10 @@ let generic = { version, sha256 }:
       ++ stdenv.lib.optionals (!isPy3k) [ funcsigs ]
       ++ stdenv.lib.optionals (pythonOlder "3.6") [ pathlib2 ];
 
-    checkPhase = ''
+    checkPhase = if isPyPy then ''
```
Comment (veprbl) [inline pkgs/development/python-modules/pytest/default.nix]: `doCheck = !isPyPy`?

File: pkgs/development/python-modules/funcsigs/default.nix (hunk 1 above)
Comment (veprbl) [reply pkgs/development/python-modules/funcsigs/default.nix]: There is a section on this in the manual: https://nixos.org/nixpkgs/manual/#sec-patches

File: pkgs/development/python-modules/pytest/default.nix (hunk 2 above)
Comment (leo60228) [reply pkgs/development/python-modules/pytest/default.nix]: I wasn't sure if the hooks were necessary.

**Comment by leo60228** [comment]
I've used fetchpatch and doCheck, currently building. My VPS only has 1GB of RAM, so this is pretty slow.

**Comment by domenkozar** [comment]
CI seems to fail: https://github.com/NixOS/nixpkgs/pull/53771/checks?check_run_id=48590077

**Comment by leo60228** [comment]
That's an upstream bug. Will fix.

**Comment by leo60228** [comment]
I've disabled the tests (which has been done on mock master, but not in a release yet).

**Comment by leo60228** [comment]
Also, is there a good way to get an armv7 PyPy in nixpkgs? PyPy has a very long compilation (>1 hour and 4GB of RAM), and I believe (though I may be wrong) that cross-compilation changes the hash. In my fork I've added a derivation based off the official binaries, but that seems like a hack. PyPy doesn't support aarch64.

**Comment by domenkozar** [comment]
Sadly there's no provider with armv7 that would be able to support our farm :)

**Comment by leo60228** [comment]
I've fixed the issues so far.

**Comment by leo60228** [comment]
@GrahamcOfBorg build pypyPackages.numpy

**Comment by domenkozar** [comment]
@GrahamcOfBorg build pypyPackages.numpy

**Comment by leo60228** [comment]
Hm. What's the error? I don't see one.

**Review by FRidh — CHANGES_REQUESTED** [review]
This needs to target `staging`.

**Comment by FRidh** [comment]
Add a comment when disabling tests.

**Comment by flokli** [comment]
What's the status of this? Changes should have landed in master by now…

**Comment by leo60228** [comment]
@flokli Status is that I forgot about it and am currently making the requested changes.

**Comment by flokli** [comment]
Cool, thanks!

**Comment by leo60228** [comment]
Closing in favor of a new pull request against staging.

**Comment by leo60228** [comment]
#64200

**Pull request closed without merging**